File: app/src/main/java/nl/rsdt/japp/jotial/maps/kml/KmlDocument.java
package nl.rsdt.japp.jotial.maps.kml;

import java.util.LinkedList;
import java.util.List;


class KmlDocument {


    private String name;
    private List<KmlFolder> folders = new LinkedList<>();
    private List<KmlStyle> styles = new LinkedList<>();

    public void setName(String name) {
        this.name = name;
    }

    public void addFolder(KmlFolder folder) {
        this.folders.add(folder);
    }

    public void addStyle(KmlStyle style) {
        this.styles.add(style);
    }

    public KmlFile toKmlFile() {
        KmlScoutingGroep organisatie = null;
        List<KmlScoutingGroep> groepen = new LinkedList<>();
        List<KmlStyleBase> styles = new LinkedList<>();
        List<KmlDeelgebied> deelgebieden = new LinkedList<>();
        for (KmlFolder folder : folders){
            switch (folder.getType()){
                case Organisatie:
                    organisatie = folder.toKmlOrganisatie();
                    break;
                case Groepen:
                    groepen = folder.toKmlGroepen();
                    break;
                case Deelgebieden:
                    deelgebieden = folder.toKmlDeelgebieden();
                    break;
            }
        }
        for (KmlStyle style : this.styles){
            KmlStyleBase styleBase = style.toStyleBase();
            if (styleBase instanceof KmlGroepStyle) {
                KmlGroepStyle groepStyle = (KmlGroepStyle) styleBase;
                for (KmlScoutingGroep groep : groepen) {
                    if (groep.getStyleUrl().equals(groepStyle.getId())) {
                        groep.setStyle(groepStyle);
                    }
                }
            }
            if (styleBase instanceof KmlDeelgebiedStyle) {
                KmlDeelgebiedStyle deelgebiedStyle = (KmlDeelgebiedStyle) styleBase;
                for (KmlDeelgebied deelgebied : deelgebieden) {
                    if (deelgebied.getStyleId().equals(deelgebiedStyle.getId())) {
                        deelgebied.setStyle(deelgebiedStyle);
                    }
                }
            }
        }

        return new KmlFile(this.name,organisatie, groepen, styles, deelgebieden);
    }
}

File: app/src/main/java/nl/rsdt/japp/jotial/maps/kml/KmlDocument.kt
package nl.rsdt.japp.jotial.maps.kml

import java.util.*


internal class KmlDocument {


    private var name: String? = null
    private val folders = LinkedList<KmlFolder>()
    private val styles = LinkedList<KmlStyle>()

    fun setName(name: String) {
        this.name = name
    }

    fun addFolder(folder: KmlFolder) {
        this.folders.add(folder)
    }

    fun addStyle(style: KmlStyle) {
        this.styles.add(style)
    }

    fun toKmlFile(): KmlFile {
        var organisatie: KmlScoutingGroep? = null
        var groepen: List<KmlScoutingGroep> = LinkedList()
        val styles = LinkedList<KmlStyleBase>()
        var deelgebieden: List<KmlDeelgebied> = LinkedList()
        for (folder in folders) {
            when (folder.type) {
                KmlFolder.Type.Organisatie -> organisatie = folder.toKmlOrganisatie()
                KmlFolder.Type.Groepen -> groepen = folder.toKmlGroepen()
                KmlFolder.Type.Deelgebieden -> deelgebieden = folder.toKmlDeelgebieden()
            }
        }
        for (style in this.styles) {
            val styleBase = style.toStyleBase()
            if (styleBase is KmlGroepStyle) {
                for (groep in groepen) {
                    if (groep.styleUrl == styleBase.id) {
                        groep.style = styleBase
                    }
                }
            }
            if (styleBase is KmlDeelgebiedStyle) {
                for (deelgebied in deelgebieden) {
                    if (deelgebied.styleId == styleBase.id) {
                        deelgebied.style = styleBase
                    }
                }
            }
        }

        return KmlFile(this.name, organisatie, groepen, styles, deelgebieden)
    }
}
